admin/sql/CreateFunctions.sql

```diff
@@ -1255,6 +1255,86 @@ BEGIN
 END;
 $$ LANGUAGE 'plpgsql';
 
+-------------------------------------------------------------------
+-- Ratings
+-------------------------------------------------------------------
+
+CREATE OR REPLACE FUNCTION update_aggregate_rating(entity_type ratable_entity_type, entity_id INTEGER)
+RETURNS VOID AS $$
+BEGIN
+  -- update the aggregate rating for the given entity_id.
+  EXECUTE format(
+    $SQL$
+      UPDATE %2$I
+         SET rating = agg.rating,
+             rating_count = nullif(agg.rating_count, 0)
+        FROM (
+          SELECT count(rating)::INTEGER AS rating_count,
+                 -- trunc(x + 0.5) is used because round() on REAL values
+                 -- rounds to the nearest even number.
+                 trunc((sum(rating)::REAL /
+                        count(rating)::REAL) +
+                       0.5::REAL)::SMALLINT AS rating
+            FROM %3$I
+           WHERE %1$I = $1
+        ) agg
+       WHERE id = $1
+    $SQL$,
+    entity_type::TEXT,
+    entity_type::TEXT || '_meta',
+    entity_type::TEXT || '_rating_raw'
+  ) USING entity_id;
+END;
+$$ LANGUAGE 'plpgsql';
+
+CREATE OR REPLACE FUNCTION update_aggregate_rating_for_raw_insert()
+RETURNS trigger AS $$
+DECLARE
+  entity_type ratable_entity_type;
+  new_entity_id INTEGER;
+BEGIN
+  entity_type := TG_ARGV[0]::ratable_entity_type;
+  EXECUTE format('SELECT ($1).%s', entity_type::TEXT) INTO new_entity_id USING NEW;
+  PERFORM update_aggregate_rating(entity_type, new_entity_id);
+  RETURN NULL;
+END;
+$$ LANGUAGE 'plpgsql';
+
+CREATE OR REPLACE FUNCTION update_aggregate_rating_for_raw_update()
+RETURNS trigger AS $$
+DECLARE
+  entity_type ratable_entity_type;
+  new_entity_id INTEGER;
+  old_entity_id INTEGER;
+BEGIN
+  entity_type := TG_ARGV[0]::ratable_entity_type;
+  EXECUTE format('SELECT ($1).%s', entity_type) INTO new_entity_id USING NEW;
+  EXECUTE format('SELECT ($1).%s', entity_type) INTO old_entity_id USING OLD;
+  IF (old_entity_id = new_entity_id AND OLD.rating != NEW.rating) THEN
+    -- Case 1: only the rating changed.
+    PERFORM update_aggregate_rating(entity_type, old_entity_id);
+  ELSIF (old_entity_id != new_entity_id OR OLD.rating != NEW.rating) THEN
+    -- Case 2: the entity or rating changed.
+    PERFORM update_aggregate_rating(entity_type, old_entity_id);
+    PERFORM update_aggregate_rating(entity_type, new_entity_id);
```

Comment:
Do we ever change the entity or is this just in case? Not that I think it's bad to have, just trying to make sure.

Reply:
I believe only for merges.